src/models/nulog/NuLogParser.py

```diff
@@ -0,0 +1,516 @@
+        )
+
+    def load_model(self, model, model_opt):
+        if using_GPU:
+            ckpt = torch.load(self.model_path)
+        else:
+            ckpt = torch.load(self.model_path, map_location=torch.device("cpu"))
+        try:
+            model_opt.load_state_dict(ckpt["optimizer_state_dict"])
+            model.load_state_dict(ckpt["model_state_dict"])
+            epoch = ckpt["epoch"]
+            loss = ckpt["loss"]
+            return epoch, loss
+        except:  ## TODO: remove this try except when we use the new save function.
+            logging.error("loading trained model with old format.")
+            model.load_state_dict(ckpt)
+
+    def train(
+        self,
+        data_tokenized,
+        batch_size=32,
+        mask_percentage=1.0,
+        pad_len=64,
+        N=1,
+        d_model=256,
+        dropout=0.1,
+        lr=0.001,
+        betas=(0.9, 0.999),
+        weight_decay=0.005,
+        nr_epochs=5,
+        num_samples=0,
+        step_size=100,
+    ):
+        self.mask_percentage = mask_percentage
+        self.pad_len = pad_len
+        self.batch_size = batch_size
+        self.N = N
+        self.d_model = d_model
+        self.dropout = dropout
+        self.lr = lr
+        self.betas = betas
+        self.weight_decay = weight_decay
+        self.num_samples = num_samples
+        self.nr_epochs = nr_epochs
+        self.step_size = step_size
+
+        logging.debug("learning rate : " + str(self.lr))
+        # test!
+        # data_tokenized = data_tokenized[:10]
+        transform_to_tensor = transforms.Lambda(lambda lst: torch.tensor(lst))
+
+        criterion = nn.CrossEntropyLoss()
+        model = self.make_model(
+            self.tokenizer.n_words,
+            self.tokenizer.n_words,
+            N=self.N,
+            d_model=self.d_model,
+            d_ff=self.d_model,
+            dropout=self.dropout,
+            max_len=self.pad_len,
+        )
+
+        if using_GPU:
+            model.cuda()
+        model_opt = torch.optim.Adam(
+            model.parameters(),
+            lr=self.lr,
+            betas=self.betas,
+            weight_decay=self.weight_decay,
+        )
+
+        if (self.logName + ".pt") in os.listdir(self.savePath):
+            # model.load_state_dict(torch.load(self.model_path))
+            prev_epoch, prev_loss = self.load_model(model, model_opt)
+
+        train_dataloader = self.get_train_dataloaders(
+            data_tokenized, transform_to_tensor
+        )
+        ## train if no model
+        model.train()
+        logging.info("#######Training Model within {self.nr_epochs} epochs...######")
+        for epoch in range(self.nr_epochs):
+            logging.info("Epoch: {}".format(epoch))
+            self.run_epoch(
+                train_dataloader,
+                model,
+                SimpleLossCompute(model.generator, criterion, model_opt),
+            )
+
+        # self.save_model(model, model_opt, epoch, 0)
```

Comment:
Should self.save_model be used? Additionally, if it used we should not use 0 for the epoch but rather self.nr_epochs

Reply:
yes we should use it. I changed it to self.nr_epochs as you suggested.